certutils/reloader.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//     http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This code is based on the example in https://github.com/kelseyhightower/kube-cert-manager/blob/master/tls-app/certificate-manager.go
+
+package certutils
+
+import (
+	"crypto/tls"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+// CertReloader provides a mechanism for reloading a TLS key and cert upon file change
+type CertReloader struct {
+	sync.RWMutex
+	certificate *tls.Certificate
+	certFile    string
+	keyFile     string
+	Error       chan error
+	watcher     *fsnotify.Watcher
+}
+
+// NewCertReloader returns a new CertReloader
+func NewCertReloader(certFile, keyFile string) (*CertReloader, error) {
+	cr := &CertReloader{
+		certFile: certFile,
+		keyFile:  keyFile,
+		Error:    make(chan error, 10),
```

Comment:
The `watchCertificate()` goroutine will stop processing when this buffer fills.  Perhaps we should add a timeout to sending on this channel so we drop errors when the user of the library fails to read from this channel?

Reply:
In general I am super uncomfortable with the Error channel. I spent half my labor day weekend trying to think of a better way to handle this. I would prefer to find a way to make the reloader uber simple and uber transparent to callers. That is, I want it to be just as easy as using TLS with the standard "load once" approach.

I'm open to suggestions on how to simplify the error handling.. some current thoughts, though not super happy about any of them:

1. just fatal out on errors.. but having a library exit your app is no good.
2. just log the errors ... optionally allow the user to supply a logger, else use a default logger
3. continue to provide the Error channel for callers that want to implement their own handling of errors. For callers that don't consume the error channel, setup a go routine that simply logs errors for them.
4. no channels, but force the caller to provide a callback func to execute on any errors.

all of these have some kind of additional problem/tradeoffs though.

thoughts? /cc @spheromak too

Reply:
# 1 has another option:
Rather than hiding the watcher goroutine from the user, require the user to call itself.  Then the watcher goroutine could bubble errors back to the user of the library merely by exiting (and leave it up to the user of the library how they want to handle that).

Other functions in the stdlib follow a similar pattern to the above (e.g. `http.ListenAndServe`).

# 2 and 4 are similar, and either would be fine IMHO.  Slight preference for the callback though as I think it offers some implicit intuitive flexibility (e.g. only call the callback if callback exists, otherwise ignore the error).  This pattern is found in the stdlib too (it's what we're implementing in this PR with `tls.Config`)

# 3 would be a little quirky IMHO.  I think I would expect a library I use not to log unless I ask it to.

Reply:
tnx @kf6nux , good ideas. I'll probably not be able to revisit this until the weekend. When I do i think i will explore your suggestion #1 first.